src/transport/iroh_net.rs

```diff
@@ -103,40 +113,97 @@ impl<In: RpcMessage, Out: RpcMessage> IrohNetServerEndpoint<In, Out> {
+            let connection = match connecting.await {
+                Ok(connection) => connection,
                 Err(e) => {
                     tracing::warn!("Error accepting connection: {}", e);
                     continue;
                 }
             };
+
+            // When the `allowed_node_ids` is empty, it's empty forever, so the CPU's branch
+            // prediction should always optimize this block away from this loop.
+            // The same applies when it isn't empty, ignoring the check for emptiness and always
+            // extracting the node id and checking if it's in the set.
+            if !allowed_node_ids.is_empty() {
+                let Ok(client_node_id) = iroh_net::endpoint::get_remote_node_id(&connection)
+                    .map_err(|e| {
+                        tracing::error!(
+                            ?e,
+                            "Failed to extract iroh-net node id from incoming connection from {:?}",
+                            connection.remote_address()
+                        )
+                    })
+                else {
+                    connection.close(0u32.into(), b"failed to extract iroh-net node id");
+                    continue;
+                };
+
+                if !allowed_node_ids.contains(&client_node_id) {
+                    connection.close(0u32.into(), b"forbidden node id");
+                    continue;
+                }
+            }
+
             tracing::debug!(
                 "Connection established from {:?}",
-                conection.remote_address()
+                connection.remote_address()
             );
+
             tracing::debug!("Spawning connection handler...");
-            tokio::spawn(Self::connection_handler(conection, sender.clone()));
+            tokio::spawn(Self::connection_handler(connection, sender.clone()));
         }
     }
 
-    /// Create a new server channel, given a quinn endpoint.
-    ///
-    /// The endpoint must be a server endpoint.
+    /// Create a new server channel, given a quinn endpoint, with unrestricted access by node id
     ///
     /// The server channel will take care of listening on the endpoint and spawning
     /// handlers for new connections.
     pub fn new(endpoint: iroh_net::Endpoint) -> io::Result<Self> {
+        Self::new_with_access_control(endpoint, AccessControl::Unrestricted)
+    }
+
+    /// Create a new server endpoint, with specified access control
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new_with_access_control(
+        endpoint: iroh_net::Endpoint,
+        access_control: AccessControl,
+    ) -> io::Result<Self> {
+        let allowed_node_ids = match access_control {
+            AccessControl::Unrestricted => BTreeSet::new(),
+            AccessControl::Allowed(list) if list.is_empty() => {
+                tracing::warn!(
```

Comment:
I think empty should mean nobody has access. E.g. you got a config file where you define allowed node ids, and you set this to `[]` accidentally or because you want to remove access. This should not make it unrestricted.

Reply:
Yeah, makes sense indeed, I will remove this `warn!` and return an error instead, better fail early than being unresponsive for all requests